Hypercube Sampling Bounds

samuel-hg-engel opened the pull request:
Hypercube sampling now supports sampling between user specified bounds, or if not specified between 0-1.

samuel-hg-engel committed:
Implemented parameter ranges to hypercube sampling

Hypercube sampling now supports sampling between user specified bounds, or if not specified between 0-1.

hpcflow-actions committed:
pre-commit fixes [skip ci]

aplowman commented on hpcflow/sdk/core/parameters.py:
```diff
@@ -1583,6 +1583,7 @@ def _values_from_latin_hypercube(
         cls,
         paths: Sequence[str],
         num_samples: int,
+        bounds: dict[str, Sequence[float]],
```
I think we want `bounds` to be optional, so could you change this to a keyword-only argument (i.e. move it after the `*` delimiter), and add a default value of `None` to it.

aplowman replied:
The type will also need to be updated to: `dict[str, Sequence[float]] | None`

aplowman commented on hpcflow/sdk/core/parameters.py:
```diff
@@ -1598,19 +1599,31 @@ def _values_from_latin_hypercube(
             optimization=optimization,
             rng=rng,
         )
+
```
...And then set bounds to an empty dict, if it is `None`: `bounds = bounds or {}`

samuel-hg-engel replied:
For this do you mean to just add this following block underneath?
```       
if bounds == None:
            bounds = {}
```

aplowman replied:
Yeah, but what you wrote is equaivalent to `bounds = bounds or {}` in this case. If bounds is `None`, we have `bounds = None or {}`, and because `None` is "falsey", the `or` will select the empty dict, so we get `bounds = bounds or {}` reducing to `bounds = {}`, in the case the function argument `bounds` is not provided.

samuel-hg-engel replied:
Oh nice, that's a much cleaner way of doing that.

aplowman commented on hpcflow/sdk/core/parameters.py:
```diff
@@ -1598,19 +1599,31 @@ def _values_from_latin_hypercube(
         )
+
+        parameter_ranges = np.array(
+            [bounds[path] if path in bounds else [0, 1] for path in paths]
```
```suggestion
            [bounds.get(path, [0, 1]) for path in paths]
```
is a bit tidier!

aplowman commented on hpcflow/sdk/core/parameters.py:
```diff
@@ -1598,19 +1599,31 @@ def _values_from_latin_hypercube(
         cls,
         paths: Sequence[str],
         num_samples: int,
+        bounds: dict[str, Sequence[float]],
```
Similarly, add a default value of `None` to `bounds` here, and make it a keyword-only argument.

aplowman reviewed (changes requested):
Looks good, thanks @samuel-hg-engel. A few refinements to make sure `bounds` is optional, which I think is the intention. These changes should then fix the failing unit tests (e.g.: https://github.com/hpcflow/hpcflow-new/actions/runs/14592463693/job/40930696533).

Please could you also add a unit test that checks the scaling is as expected? You can do something like: 

```python
bounds = {...} # some values far away from [0, 1]

mps = hf.MultiPathSequence.from_latin_hypercube(
    paths=["inputs.a"],
    num_samples=10,
    bounds=bounds,
)

vals_arr = np.array(mps.values)
assert vals_arr.shape == (1, 10)

vals_0 = vals_arr[0] 
bounds_0 = list(bounds.values())[0]
assert np.logical_and(vals_0 > bounds_0[0], vals_0 < bounds_0[1]).all()
```

samuel-hg-engel committed:
Added changes to parameter.py for pull request

Bounds is now an optional argument, which default to the standard hypercube ranges.

samuel-hg-engel committed:
Update parameters.py

hpcflow-actions committed:
pre-commit fixes [skip ci]

samuel-hg-engel committed:
Update test_multi_path_sequences.py

Added unit test for the bounds of the latin hypercube sampler. It checks that the samples are within the user specified bound, additionally checks for samples which do not have a specified bounds.

samuel-hg-engel committed:
Merge branch 'calibration-workflow' of https://github.com/hpcflow/hpcflow-new into calibration-workflow

hpcflow-actions committed:
pre-commit fixes [skip ci]

aplowman reviewed (approved):
Excellent!

aplowman commented:
This fixed #790
